File: server/src/utils/rateLimiter.ts
/**
 * Simple rate limiter for email sending
 * Ensures emails are sent at controlled intervals to prevent overwhelming email providers
 */

export interface RateLimitConfig {
  intervalMs: number;
  maxPerMinute: number;
  maxPerHour: number;
}

export class EmailRateLimiter {
  private lastSendTime: number = 0;
  private sendCount: number = 0;
  private minuteResetTime: number = 0;
  private hourResetTime: number = 0;
  private minuteCount: number = 0;
  private hourCount: number = 0;

  constructor(private config: RateLimitConfig) {}

  /**
   * Wait until it's safe to send the next email
   */
  async waitForNextSend(): Promise<void> {
    const now = Date.now();

    // Reset counters if time windows have passed
    if (now >= this.minuteResetTime) {
      this.minuteCount = 0;
      this.minuteResetTime = now + 60 * 1000; // Next minute
    }

    if (now >= this.hourResetTime) {
      this.hourCount = 0;
      this.hourResetTime = now + 60 * 60 * 1000; // Next hour
    }

    // Check rate limits
    if (this.minuteCount >= this.config.maxPerMinute) {
      const waitTime = this.minuteResetTime - now;
      if (waitTime > 0) {
        await this.delay(waitTime);
        return this.waitForNextSend(); // Recursively check again after waiting
      }
    }

    if (this.hourCount >= this.config.maxPerHour) {
      const waitTime = this.hourResetTime - now;
      if (waitTime > 0) {
        await this.delay(waitTime);
        return this.waitForNextSend(); // Recursively check again after waiting
      }
    }

    // Check time-based interval
    const timeSinceLastSend = now - this.lastSendTime;
    const requiredWait = this.config.intervalMs - timeSinceLastSend;

    if (requiredWait > 0) {
      await this.delay(requiredWait);
    }

    // Update counters and timestamps
    this.lastSendTime = Date.now();
    this.sendCount++;
    this.minuteCount++;
    this.hourCount++;
  }

  /**
   * Create a delay using setTimeout
   */
  private delay(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Get current statistics (for monitoring/debugging)
   */
  getStats() {
    const now = Date.now();
    return {
      lastSendTime: this.lastSendTime,
      sendCount: this.sendCount,
      minuteCount: this.minuteCount,
      hourCount: this.hourCount,
      timeToNextMinute: Math.max(0, this.minuteResetTime - now),
      timeToNextHour: Math.max(0, this.hourResetTime - now),
      config: this.config
    };
  }

  /**
   * Reset the rate limiter (useful for testing or recovery)
   */
  reset(): void {
    this.lastSendTime = 0;
    this.sendCount = 0;
    this.minuteResetTime = 0;
    this.hourResetTime = 0;
    this.minuteCount = 0;
    this.hourCount = 0;
  }
}

// Global rate limiter instance
let globalRateLimiter: EmailRateLimiter | null = null;

/**
 * Get or create the global rate limiter instance
 */
export function getGlobalRateLimiter(): EmailRateLimiter {
  if (!globalRateLimiter) {
    const { emailRateLimit } = require('../config/env').env;
    globalRateLimiter = new EmailRateLimiter(emailRateLimit);
  }
  return globalRateLimiter;
}

/**
 * Update the rate limiter configuration at runtime
 */
export function updateRateLimiterConfig(newConfig: Partial<RateLimitConfig>): void {
  const limiter = getGlobalRateLimiter();
  const currentConfig = limiter.getStats().config;
  
  // Update the configuration
  const updatedConfig = { ...currentConfig, ...newConfig };
  
  // Create a new limiter with the updated config
  globalRateLimiter = new EmailRateLimiter(updatedConfig);
}

/**
 * Get current rate limiter configuration
 */
export function getCurrentRateLimitConfig(): RateLimitConfig {
  const limiter = getGlobalRateLimiter();
  return limiter.getStats().config;
}

/**
 * Convenience function to wait for next email send
 */
export async function waitForNextEmailSend(): Promise<void> {
  const limiter = getGlobalRateLimiter();
  await limiter.waitForNextSend();
}

/**
 * Get current rate limiter statistics
 */
export function getRateLimiterStats() {
  const limiter = getGlobalRateLimiter();
  return limiter.getStats();
}

File: backend/app/utils/rate_limiter.py
import asyncio
import time
from typing import Dict, Any, Optional
from datetime import datetime, timedelta
from ..models.rate_limit import RateLimitConfig, RateLimitStats
from ..config import get_settings


class EmailRateLimiter:
    def __init__(self, config: RateLimitConfig):
        self.config = config
        self.last_send_time: int = 0
        self.send_count: int = 0
        self.minute_reset_time: int = 0
        self.hour_reset_time: int = 0
        self.minute_count: int = 0
        self.hour_count: int = 0
    
    async def wait_for_next_send(self) -> None:
        now = int(time.time() * 1000)  # Convert to milliseconds
        
        # Reset counters if time windows have passed
        if now >= self.minute_reset_time:
            self.minute_count = 0
            self.minute_reset_time = now + 60 * 1000  # Next minute
        
        if now >= self.hour_reset_time:
            self.hour_count = 0
            self.hour_reset_time = now + 60 * 60 * 1000  # Next hour
        
        # Check rate limits
        if self.minute_count >= self.config.max_per_minute:
            wait_time = self.minute_reset_time - now
            if wait_time > 0:
                await asyncio.sleep(wait_time / 1000)  # Convert to seconds
                return await self.wait_for_next_send()  # Recursively check again
        
        if self.hour_count >= self.config.max_per_hour:
            wait_time = self.hour_reset_time - now
            if wait_time > 0:
                await asyncio.sleep(wait_time / 1000)  # Convert to seconds
                return await self.wait_for_next_send()  # Recursively check again
        
        # Check time-based interval
        time_since_last_send = now - self.last_send_time
        required_wait = self.config.interval_ms - time_since_last_send
        
        if required_wait > 0:
            await asyncio.sleep(required_wait / 1000)  # Convert to seconds
        
        # Update counters and timestamps
        self.last_send_time = int(time.time() * 1000)
        self.send_count += 1
        self.minute_count += 1
        self.hour_count += 1
    
    def get_stats(self) -> RateLimitStats:
        now = int(time.time() * 1000)
        return RateLimitStats(
            last_send_time=self.last_send_time,
            send_count=self.send_count,
            minute_count=self.minute_count,
            hour_count=self.hour_count,
            time_to_next_minute=max(0, self.minute_reset_time - now),
            time_to_next_hour=max(0, self.hour_reset_time - now),
            config=self.config
        )
    
    def reset(self) -> None:
        self.last_send_time = 0
        self.send_count = 0
        self.minute_reset_time = 0
        self.hour_reset_time = 0
        self.minute_count = 0
        self.hour_count = 0


# Global rate limiter instance
_global_rate_limiter: Optional[EmailRateLimiter] = None


def get_global_rate_limiter() -> EmailRateLimiter:
    global _global_rate_limiter
    if _global_rate_limiter is None:
        settings = get_settings()
        _global_rate_limiter = EmailRateLimiter(RateLimitConfig(**settings.email_rate_limit))
    return _global_rate_limiter


def update_rate_limiter_config(new_config: Dict[str, Any]) -> None:
    global _global_rate_limiter
    limiter = get_global_rate_limiter()
    current_config = limiter.get_stats().config
    
    # Update the configuration
    updated_config = {**current_config.dict(), **new_config}
    
    # Create a new limiter with the updated config
    _global_rate_limiter = EmailRateLimiter(RateLimitConfig(**updated_config))


def get_current_rate_limit_config() -> RateLimitConfig:
    limiter = get_global_rate_limiter()
    return limiter.get_stats().config


async def wait_for_next_email_send() -> None:
    limiter = get_global_rate_limiter()
    await limiter.wait_for_next_send()


def get_rate_limiter_stats() -> RateLimitStats:
    limiter = get_global_rate_limiter()
    return limiter.get_stats()